### PR TITLE
Usage rights properties

### PR DESCRIPTION
**Pull request by jamesgorrie**
Creating forms somewhat dynamically.
Seems a little like overkill - but a lot simpler than having all the logic on the client.

I want to look into reflection similar to the Play JSON stuff or perhaps a lib, but I can't find one that could do the nice stuff like the optionsMap I have here.

usage-rights-staff-phot

### CODE REVIEW

File: kahuna/public/js/usage-rights/usage-rights-editor.html [hunk 1]
```diff
@@ -0,0 +1,82 @@
+<form class="usage-rights-editor ure"
```
Comment (kenoir) [inline kahuna/public/js/usage-rights/usage-rights-editor.html]: Can you remove the `ure` class here?

File: metadata-editor/app/controllers/EditsApi.scala [hunk 2]
```diff
@@ -42,7 +43,8 @@ object EditsApi extends Controller with ArgoHelpers {
     // their images can be imported by drag and drop instead
     // FIXME: Creating new instances? Rubbish ಠ_ಠ.
     val usageRightsData =
-      List(PrImage(), Handout(), Screengrab(), SocialMedia(), Obituary(), Pool())
+      List(PrImage(), Handout(), Screengrab(), SocialMedia(), Obituary(), Pool(),
+           StaffPhotographer("?", "?"))
```
Comment (theefer) [inline metadata-editor/app/controllers/EditsApi.scala]: Sorry I know I asked before, but is there no better way to do this? If this is really the only way, can we document?

**Comment by theefer** [comment]
Could you post some example JSON that the API would return?

**Comment by jamesgorrie** [comment]
```JSON
{
    "data": [
        {
            "value": "PR Image",
            "name": "PR Image",
            "cost": "conditional",
            "description": "Used to promote specific exhibitions, auctions, etc. and only available for such purposes.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": true
                }
            ]
        },
        {
            "value": "handout",
            "name": "Handout",
            "cost": "free",
            "description": "Provided free of use for press purposes e.g. police images for new stories, family shots in biographical pieces, etc.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": false
                }
            ]
        },
        {
            "value": "screengrab",
            "name": "Screengrab",
            "cost": "conditional",
            "description": "Still images created by us from moving footage in television broadcasts usually in relation to breaking news stories.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": true
                }
            ]
        },
        {
            "value": "social-media",
            "name": "Social Media",
            "cost": "conditional",
            "description": "Images taken from public websites and social media to support breaking news where no other image is available from usual sources. Permission should be sought from the copyright holder, but in extreme circumstances an image may be used with the approval of a senior editor.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": true
                }
            ]
        },
        {
            "value": "obituary",
            "name": "Obituary",
            "cost": "conditional",
            "description": "Acquired from private sources, e.g. family members, for the purposes of obituaries.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": true
                }
            ]
        },
        {
            "value": "pool",
            "name": "Pool",
            "cost": "conditional",
            "description": "Images issued during major national events that are free to use andshared amongst news media organisations during that event. Rights revert to the copyright holder when the pool is terminated.",
            "properties": [
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": true
                }
            ]
        },
        {
            "value": "staff-photographer",
            "name": "Staff Photographer",
            "cost": "free",
            "description": "Pictures created by photographers who are or were members of staff.",
            "properties": [
                {
                    "name": "publication",
                    "label": "Publication",
                    "type": "string",
                    "required": true,
                    "options": [
                        "The Guardian",
                        "The Observer"
                    ]
                },
                {
                    "name": "photographer",
                    "label": "Photographer",
                    "type": "string",
                    "required": true,
                    "optionsMap": {
                        "The Guardian": [
                            "Graeme Robertson",
                            "Mike Bowers",
                            "Christopher Thomond",
                            "Martin Argles",
                            "Tom Jenkins",
                            "Linda Nylind",
                            "Eamonn Mccabe",
                            "Tristram Kenton",
                            "Sean Smith",
                            "Frank Baron",
                            "David Sillitoe",
                            "Sarah Lee",
                            "Murdo Macleod",
                            "Graham Turner",
                            "Martin Godwin",
                            "David Levene",
                            "Felicity Cloake"
                        ],
                        "The Observer": [
                            "Andy Hall",
                            "Suki Dhanda",
                            "Gary Calton",
                            "Sophia Evans",
                            "Karen Robinson",
                            "Catherine Shaw",
                            "Katherine Anne Rose",
                            "Antonio Olmos",
                            "Richard Saker"
                        ]
                    },
                    "optionsMapKey": "publication"
                },
                {
                    "name": "restrictions",
                    "label": "Restrictions",
                    "type": "text",
                    "required": false
                }
            ]
        }
    ]
}```

File: metadata-editor/app/controllers/EditsApi.scala (hunk 2 above)
Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsApi.scala]: I could have something like this:
``` Scala
case class BigBang(restrictions: Option[String]) extends UsageRights {
  val category = BigBang.category
  val defaultCost = BigBang.defaultCost
  val description = BigBang.description
}
object BigBang {
  val category = "big-bang"
  val defaultCost = Some(Conditional)
  val description =
    "Used to promote specific exhibitions, auctions, etc. and only available " +
    "for such purposes."
}
```

What do you think?

Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsApi.scala]: You;d actually have to do something like:
``` Scala
sealed trait UsageRightsProps {
  val category: String
  val defaultCost: Option[Cost]
  val description: String
}

case class BigBang(restrictions: Option[String]) extends UsageRights {
  val category = BigBang.category
  val defaultCost = BigBang.defaultCost
  val description = BigBang.description
}
object BigBang extends UsageRightsProps {
  val category = "big-bang"
  val defaultCost = Some(Conditional)
  val description =
    "Used to promote specific exhibitions, auctions, etc. and only available " +
    "for such purposes."
}

case class BigBang2(restrictions: Option[String]) extends UsageRights {
  val category = BigBang.category
  val defaultCost = BigBang.defaultCost
  val description = BigBang.description
}
object BigBang2 extends UsageRightsProps {
  val category = "big-bang2"
  val defaultCost = Some(Conditional)
  val description =
    "Used to promote specific exhibitions, auctions, etc. and only available " +
    "for such purposes."
}



object test {
  BigBang.category

  val t = List(BigBang, BigBang2).map(_.category)
}
```

Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsApi.scala]: I've added a note explaining why I do that - I'll keep searching for a better way as I'm not massively happy with the above examples.

File: metadata-editor/app/controllers/EditsApi.scala [hunk 3]
```diff
@@ -40,9 +41,11 @@ object EditsApi extends Controller with ArgoHelpers {
   val usageRightsResponse = {
     // FIXME: GuardianWitness should be there but isn't for simplicity;
     // their images can be imported by drag and drop instead
-    // FIXME: Creating new instances? Rubbish ಠ_ಠ.
+    // FIXME: Creating new instances? Rubbish ಠ_ಠ. I can't think of a way
+    // to access the `val`s of the classes though without instantiating them.
     val usageRightsData =
-      List(PrImage(), Handout(), Screengrab(), SocialMedia(), Obituary(), Pool())
+      List(PrImage(), Handout(), Screengrab(), SocialMedia(), Obituary(), Pool(),
```
Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsApi.scala]: @theefer would you like me to spend more time at making this better?
I don't particularly like it either, but feel I could be building the interface first, so long as the shape of the API is correct.

Comment (theefer) [inline metadata-editor/app/controllers/EditsApi.scala]: No the comment is fine.

**Comment by jamesgorrie** [comment]
Does someone want me to run them through this?

**Comment by kenoir** [comment]
I'm happy to today.

On Wed, 15 Jul 2015 08:03 James Gorrie <notifications@github.com> wrote:

> Does someone want me to run them through this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/grid/pull/919#issuecomment-121510996>.
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

**Comment by kenoir** [comment]
:+1:

**Pull request merged**